hypershift-operator/controllers/nodepool/kubevirt/imagecacher.go

```diff
@@ -0,0 +1,263 @@
+	if err != nil {
+		return fmt.Errorf("failed to get old PVCs; %w", err)
+	}
+	labelSelector := labels.NewSelector().Add(*roleMatcher)
+
+	pvcList := &corev1.PersistentVolumeClaimList{}
+	err = cl.List(ctx, pvcList, &client.ListOptions{
+		Namespace:     qi.namespace,
+		LabelSelector: labelSelector,
+	})
+
+	if err != nil {
+		return fmt.Errorf("failed to get old PVCs; %w", err)
+	}
+
+	if len(pvcList.Items) > 1 {
+		for _, oldPVC := range pvcList.Items {
+			if oldPVC.Name == qi.dvName {
+				continue
+			}
+
+			if oldPVC.DeletionTimestamp == nil {
+				err = cl.Delete(ctx, &oldPVC)
+				if err != nil {
+					qi.log.Error(err, fmt.Sprintf("failed to delete an old PVC; namespace: %s, name: %s", oldPVC.Namespace, oldPVC.Name))
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func (qi qcowImage) createDVForCache(ctx context.Context, cl client.Client, kvPlatform *hyperv1.KubevirtNodePoolPlatform) error {
+	dv := qi.buildDVForCache(kvPlatform)
+
+	err := cl.Create(ctx, dv)
+	if err != nil && !errors.IsAlreadyExists(err) {
+		return fmt.Errorf("failed to create a DataVolume for the boot image cache ; %w", err)
+	}
+
+	return nil
+}
+
+func (qi qcowImage) getDVSourceForVMTemplate() *v1beta1.DataVolumeSource {
+	return &v1beta1.DataVolumeSource{
+		PVC: &v1beta1.DataVolumeSourcePVC{
+			Namespace: qi.namespace,
+			Name:      qi.dvName,
+		},
+	}
+}
+
+func (qi qcowImage) buildDVForCache(kvPlatform *hyperv1.KubevirtNodePoolPlatform) *v1beta1.DataVolume {
+	var (
+		volumeSize   apiresource.Quantity
+		storageClass *string
+	)
+
+	if kvPlatform.RootVolume != nil && kvPlatform.RootVolume.Persistent != nil {
+		if kvPlatform.RootVolume.Persistent.Size != nil {
+			volumeSize = *kvPlatform.RootVolume.Persistent.Size
+		}
+		storageClass = kvPlatform.RootVolume.Persistent.StorageClass
+	}
+
+	return &v1beta1.DataVolume{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      qi.dvName,
+			Namespace: qi.namespace,
+			Labels: map[string]string{
+				bootImagePVCLabelRoleName: bootImagePVCLabelRoleValue,
+			},
+			Annotations: map[string]string{
+				bootImagePVCAnnotationHash:    qi.hash,
+				cdiImmediateBindingAnnotation: "true",
+				bootImagePVCLabelRoleName:     bootImagePVCLabelRoleValue,
+			},
+		},
+		Spec: v1beta1.DataVolumeSpec{
+			Source: &v1beta1.DataVolumeSource{
+				HTTP: &v1beta1.DataVolumeSourceHTTP{
+					URL: qi.name,
+				},
+			},
+			Preallocation: pointer.Bool(true),
+			Storage: &v1beta1.StorageSpec{
+				AccessModes: []corev1.PersistentVolumeAccessMode{corev1.ReadWriteOnce},
+				Resources: corev1.ResourceRequirements{
+					Requests: map[corev1.ResourceName]apiresource.Quantity{
+						corev1.ResourceStorage: volumeSize,
+					},
+				},
+				StorageClassName: storageClass,
+			},
+		},
+	}
+}
+
+// getDvName returns the first 63 chars of the prefix + the image hash
+func getDvName(hash string) string {
+	name := bootImageNamePrefix + hash
+	if len(name) > maxK8sNameLen {
+		name = name[:maxK8sNameLen]
+	}
+	return name
+}
```

Comment:
is there a reason we can't use `dv.ObjectMeta.GenerateName`  here? This max name len issue and the need to have unique names is why GenerateName works well for us here.

Reply:
Yes. Because CDI does not copy the DV labels to the PVC, it's hard to find the new PVC in the next reconcile loop, and I had to read all the PVCs in the namespace and loop over them and so on. It created too msny issues.